htlcnotifier/htlcnotifier.go

```diff
@@ -0,0 +1,300 @@
+)
+
+// String returns the string representation of a FailReason.
+func (r FailReason) String() string {
+	switch r {
+	case FailReasonHTLCBelowMinimum:
+		return "HTLC below minimum"
+
+	case FailReasonHTLCAboveMaximum:
+		return "HTLC above maximum"
+
+	case FailReasonExpiryTooSoon:
+		return "Expiry too soon"
+
+	case FailReasonExpiryTooFar:
+		return "Expiry too far"
+
+	case FailReasonInsufficientBandwidth:
+		return "Insufficient bandwidth"
+
+	case FailReasonNotForwarding:
+		return "Forwarding disabled"
+
+	case FailReasonUnknownLink:
+		return "Unknown link"
+
+	case FailReasonInsufficientFee:
+		return "Insufficient fees"
+
+	case FailReasonIncorrectCLTV:
+		return "Incorrect CLTV"
+
+	case FailReasonLinkNotEligible:
+		return "Ineligible for forwarding"
+
+	default:
+		return ""
+	}
+}
+
+// LinkFailEvent represents a HTLC which arrived at our node, but failed to be
+// forwarded over our outgoing channel. The reason for failure is provided by
+// a FailReason enum and the details of the intended ForwardEvent are supplied.
+type LinkFailEvent struct {
+	// FailReason is the reason that we did not forward an outgoing HTLC.
+	FailReason FailReason
+
+	// ForwardingEvent provides the details of the intended HTLC forward.
+	ForwardingEvent
+}
+
+// ForwardFailType classifies the type of failure has occurred down the line
+// from our node.
+type ForwardFailType int
+
+const (
+	// ForwardFailType indicates that the forwarding failure was
+	// encrypted, so we have no further information about the failure.
+	ForwardFailTypeOpaque ForwardFailType = iota
+
+	// ForwardFailTypeMalformedHTLC indicates that the remote peer was not
+	// capable of decoding the onion.
+	ForwardFailTypeMalformedHTLC
+
+	// ForwardFailTypePermanentChannel indicates that a permanent channel
+	// failure occurred down the line.
+	ForwardFailTypePermanentChannel
+)
+
+// String returns the string representation of a ForwardFailType.
+func (f ForwardFailType) String() string {
+	switch f {
+	case ForwardFailTypeOpaque:
+		return "Opaque forwarding error"
+
+	case ForwardFailTypeMalformedHTLC:
+		return "Malformed HTLC"
+
+	case ForwardFailTypePermanentChannel:
+		return "Permanent channel failure"
+
+	default:
+		return ""
+	}
+}
+
+// ForwardingFailEvent represents a HTLC failure which after we forwarded the
+// HTLC through our node. An error is not included in this event because errors
+// returned down the route are encrypted.
+type ForwardingFailEvent struct {
+	//TODO(carla): Add elapsed time field.
+
+	FailureType ForwardFailType
+
+	// ForwardingEvent provides the details of the HTLC forward on our node
+	// which is linked to the down the line failure.
+	ForwardingEvent
+}
+
+// SettleEvent represents a HTLC that was forwarded by our node and successfully
+// settled.
+type SettleEvent struct {
+	//TODO(carla): Add elapsed time field.
+
+	// ForwardingEvent provides the details of the HTLC forward on our node
+	// which was settled.
+	ForwardingEvent
+}
```

Comment:
Nodes forward htlcs, but - on the same channels - may also send or receive payments for which they are the endpoint. For profitability calculation that takes into account balance, those events are important too. Will they be communicated over this same notifier channel, or do they go a different way?

Reply:
From the insights side of things, I see our node's HTLCs as a separate concern to profitability. They capture "personal utility" gained from a channel, which could be another aspect to factor into scoring; routing nodes would prioritize profitability, non-routing would be more concerned with utility. 

But from a `htlcNotifier`, I think it makes sense to include them to keep the notifier generic. I'd probably go with a `our_htlc` bool in the settle/fail event types (since these HTLCs can't be forwarded/have link failures, but could also go with another event type entirely?

Reply:
I was mainly thinking that we also need "our htlcs" to be able to reconstruct the channel balance at the time of an event.